File: native/pact_consumer_nif/src/builders/message_builder.rs
use pact_consumer::builders::MessageInteractionBuilder;
use rustler::{NifResult, NifStruct, Resource, ResourceArc};
use std::sync::Mutex;

use crate::{
    impl_builder_nif,
    models::interaction::{InteractionResource, NifInteraction},
    patterns::NifJsonPattern,
};

#[derive(NifStruct)]
#[module = "MessageInteractionBuilder"]
pub struct NifMessageInteractionBuilder {
    inner: ResourceArc<MessageInteractionBuilderResource>,
}

pub struct MessageInteractionBuilderResource(Mutex<MessageInteractionBuilder>);

impl NifMessageInteractionBuilder {
    fn invoke<F, T>(&self, fun: F) -> NifResult<T>
    where
        F: FnOnce(&mut MessageInteractionBuilder) -> NifResult<T>,
    {
        let mut inner = self
            .inner
            .0
            .lock()
            .map_err(|_e| rustler::Error::RaiseAtom("invalid_message_builder_reference"))?;

        fun(&mut inner)
    }
}

impl Resource for MessageInteractionBuilderResource {}

#[rustler::nif(name = "message_builder_new")]
pub fn new(description: String) -> NifMessageInteractionBuilder {
    NifMessageInteractionBuilder {
        inner: ResourceArc::new(MessageInteractionBuilderResource(Mutex::new(
            MessageInteractionBuilder::new(description),
        ))),
    }
}

#[rustler::nif(name = "message_builder_given_with_params")]
pub fn given_with_params(
    builder: NifMessageInteractionBuilder,
    given: String,
    params: String,
) -> NifResult<NifMessageInteractionBuilder> {
    builder.invoke(|b| {
        let params: serde_json::Value = serde_json::from_str(&params)
            .map_err(|_e| rustler::Error::RaiseAtom("invalid_params"))?;

        b.given_with_params(given, &params);
        Ok(())
    })?;

    Ok(builder)
}

#[rustler::nif(name = "message_builder_build")]
pub fn build(builder: NifMessageInteractionBuilder) -> NifResult<NifInteraction> {
    let interaction = builder.invoke(|b| Ok(b.build()))?;

    Ok(NifInteraction {
        inner: ResourceArc::new(InteractionResource(Box::new(interaction))),
    })
}

impl_builder_nif!("message_builder_with_key", NifMessageInteractionBuilder, with_key(key: String));
impl_builder_nif!("message_builder_pending", NifMessageInteractionBuilder, pending(value: bool));
impl_builder_nif!("message_builder_given", NifMessageInteractionBuilder, given(value: String));
impl_builder_nif!("message_builder_comment", NifMessageInteractionBuilder, comment(value: String));
impl_builder_nif!("message_builder_test_name", NifMessageInteractionBuilder, test_name(name: String));
impl_builder_nif!("message_builder_json_body", NifMessageInteractionBuilder, json_body(body: NifJsonPattern));

File: lib/native/builders/message_builder.ex
defmodule Pact.Native.Builders.MessageBuilder do
  @moduledoc false

  alias Pact.Patterns
  alias Pact.Native.PactConsumer.Interaction

  defmacro __using__(_opts) do
    quote do
      defmodule MessageInteractionBuilder do
        @moduledoc """
        Message Interaction Builder
        """
        @enforce_keys [:inner]
        defstruct [:inner]

        @type t :: %__MODULE__{
                inner: reference()
              }
      end

      @spec message_builder_new(description :: String.t()) ::
              MessageInteractionBuilder.t()
      def message_builder_new(_description),
        do: :erlang.nif_error(:nif_not_loaded)

      @spec message_builder_with_key(buider :: MessageInteractionBuilder.t(), key :: String.t()) ::
              MessageInteractionBuilder.t()
      def message_builder_with_key(_builder, _key), do: :erlang.nif_error(:nif_not_loaded)

      @spec message_builder_pending(builder :: MessageInteractionBuilder.t(), pending :: bool()) ::
              MessageInteractionBuilder.t()
      def message_builder_pending(_builder, _pending), do: :erlang.nif_error(:nif_not_loaded)

      @spec message_builder_given(builder :: MessageInteractionBuilder.t(), given :: String.t()) ::
              MessageInteractionBuilder.t()
      def message_builder_given(_builder, _given), do: :erlang.nif_error(:nif_not_loaded)

      @spec message_builder_given_with_params(
              builder :: MessageInteractionBuilder.t(),
              given :: String.t(),
              params :: String.t()
            ) :: MessageInteractionBuilder.t()
      def message_builder_given_with_params(_builder, _given, _params),
        do: :erlang.nif_error(:nif_not_loaded)

      @spec message_builder_comment(
              builder :: MessageInteractionBuilder.t(),
              comment :: String.t()
            ) ::
              MessageInteractionBuilder.t()
      def message_builder_comment(_builder, _comment), do: :erlang.nif_error(:nif_not_loaded)

      @spec message_builder_test_name(
              builder :: MessageInteractionBuilder.t(),
              name :: String.t()
            ) ::
              MessageInteractionBuilder.t()
      def message_builder_test_name(_builder, _name), do: :erlang.nif_error(:nif_not_loaded)

      @spec message_builder_json_body(
              builder :: MessageInteractionBuilder.t(),
              body :: Patterns.json_pattern()
            ) ::
              MessageInteractionBuilder.t()
      def message_builder_json_body(_builder, _body), do: :erlang.nif_error(:nif_not_loaded)

      @spec message_builder_build(builder :: MessageInteractionBuilder.t()) :: Interaction.t()
      def message_builder_build(_builder), do: :erlang.nif_error(:nif_not_loaded)
    end
  end
end
